pkg/component/runtime/service_command.go

```diff
@@ -93,11 +98,195 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
 		log.Warnf("spec is nil, nothing to execute, binaryPath: %s", binaryPath)
 		return nil
 	}
-	return executeCommand(ctx, log, binaryPath, spec.Args, envSpecToEnv(spec.Env), spec.Timeout)
+
+	// If no initial sleep duration is specified, use default value
+	retrySleepInitDuration := spec.RetrySleepInitDuration
+	if retrySleepInitDuration == 0 {
+		retrySleepInitDuration = defaultRetrySleepInitDuration
+	}
+
+	serviceCmdRetrier.Start(
+		cmdCtx, log,
+		binaryPath, spec.Args, envSpecToEnv(spec.Env), spec.Timeout,
+		retryCtx, retrySleepInitDuration, retrySleepMaxDuration,
+	)
+	return nil
+}
+
+type cmdRetryInfo struct {
+	retryCancelFn context.CancelFunc
+	cmdCancelFn   context.CancelFunc
+	cmdDone       <-chan struct{}
+}
+
+type cmdRetrier struct {
+	mu   sync.RWMutex
+	cmds map[uint64]cmdRetryInfo
+}
+
+func (cr *cmdRetrier) Start(
+	cmdCtx context.Context, log *logger.Logger,
+	binaryPath string, args []string, env []string, timeout time.Duration,
+	retryCtx context.Context, retrySleepInitDuration time.Duration, retrySleepMaxDuration time.Duration,
+) {
+	cmdKey := cr.cmdKey(binaryPath, args, env)
+
+	// Due to infinite retries, we may still be trying to (re)execute
+	// a command from a previous call to Start(). We should first stop
+	// these retries as well as the command process.
+	cr.Stop(cmdKey, log)
+
+	// Track the command so we can cancel it and it's retries later.
+	cmdCtx, cmdCancelFn := context.WithCancel(cmdCtx)
+	retryCtx, retryCancelFn := context.WithCancel(retryCtx)
+	cmdDone := make(chan struct{}, 1)
+	cr.track(cmdKey, cmdCancelFn, retryCancelFn, cmdDone)
+
+	// Execute command with retries and exponential backoff between attempts
+	expBackoff := backoff.NewExponentialBackOff()
+	expBackoff.InitialInterval = retrySleepInitDuration
+	expBackoff.MaxInterval = retrySleepMaxDuration
+
+	backoffCtx := backoff.WithContext(expBackoff, retryCtx)
+
+	// Since we will be executing the command with infinite retries, we don't
+	// want to block.  So we execute the command with retries in its own
+	// goroutine.
+	go func() {
+		retryAttempt := 0
+
+		// Here we indefinitely retry the executeCommand call, as long as it
+		// returns a non-nil error. We will block here until executeCommand
+		// returns a nil error, indicating that the command being executed has
+		// successfully completed execution.
+		//nolint: errcheck // No point checking the error inside the goroutine.
+		backoff.RetryNotify(
+			func() error {
+				err := executeCommand(cmdCtx, log, binaryPath, args, env, timeout)
+				cmdDone <- struct{}{}
+				return err
+			},
+			backoffCtx,
+			func(err error, retryAfter time.Duration) {
+				retryAttempt++
+				log.Warnf(
+					"service command execution failed with error [%s], retrying (will be retry [%d]) after [%s]",
+					err.Error(),
+					retryAttempt,
+					retryAfter,
+				)
+				<-cmdDone
+			},
+		)
+
+		cr.untrack(cmdKey)
+	}()
+}
+
+func (cr *cmdRetrier) Stop(cmdKey uint64, log *logger.Logger) {
+	cr.mu.RLock()
```

Comment:
you're modifying cr.cmds here, should be a proper lock. 
Stop is not used elsewhere can be unexported and merged with untrack as it seems its idempotent

Reply:
Yes, good catch, thanks.  A previous version of this method was only reading from `cr.cmds`; I forgot to update the lock after that changed.

Reply:
Fixed.